core/src/apps/bitcoin/sign_tx/bitcoin.py

```diff
@@ -130,45 +148,95 @@ async def step3_confirm_tx(self) -> None:
         if fee < 0:
             self.on_negative_fee()
 
+        total = self.total_in - self.change_out
+        spending = total - self.external_in
+
         # fee > (coin.maxfee per byte * tx size)
         if fee > (self.coin.maxfee_kb / 1000) * (self.weight.get_total() / 4):
             await helpers.confirm_feeoverthreshold(fee, self.coin)
         if self.tx.lock_time > 0:
             await helpers.confirm_nondefault_locktime(self.tx.lock_time)
-        await helpers.confirm_total(self.total_in - self.change_out, fee, self.coin)
+        if total == spending:
+            await helpers.confirm_total(spending, fee, self.coin)
+        else:
+            await helpers.confirm_joint_total(spending, total, self.coin)
 
-    async def step4_serialize_inputs(self) -> None:
+    async def step4_verify_external_inputs(self) -> None:
+        # should come out the same as h_external, checked before continuing
+        h_check = self.create_hash_writer()
+
+        for i in sorted(self.external):
+            progress.advance()
+            txi = await helpers.request_tx_input(self.tx_req, i, self.coin)
+            writers.write_tx_input_check(h_check, txi)
+            prev_amount, script_pubkey = await self.get_prevtx_output(
+                txi.prev_hash, txi.prev_index
+            )
+            if prev_amount != txi.amount:
+                raise wire.DataError("Invalid amount specified")
+
+            if txi.ownership_proof:
+                if not verify_nonownership(
+                    txi.ownership_proof, script_pubkey, b"", self.keychain, self.coin
+                ):
+                    raise wire.DataError("Fake external input")
+            else:
+                verifier = SignatureVerifier(
+                    script_pubkey, txi.script_sig, txi.witness, self.coin
+                )
+
+                if not verifier.check_sighhash_type(self.get_hash_type() & 0xFF):
```

Comment:
why `& 0xFF` ? do we have negative sighashes?

Reply:
That is to get rid of the 24-bit most significant bits, which contain the fork ID. Actually, funny thing, the naming is a bit messed up, I should rename `check_sighhash_type` to `check_hash_type` and `get_hash_type` to `get_sighash_type`.

Reply:
5530aa356b48e6961296fc7d62a0766818ca2901